### PR TITLE
Add tests for removal of RPM orphans

### LINKED ISSUE #134: Test orphan remove
- [x] Remove all the orphaned contents
- [ ] Remove orphaned content by type

1. Create a repository
2. Sync a repository
3. Delete  the repository
4. List the orphaned content
5. Remove orphaned content
     a.Remove the orphaned content by type
     b. Remove all the orphaned content

### PR DESCRIPTION
**Pull request by asmacdo**
These tests create and sync a repository, pulling down all content units
for that repository. The repo is deleted leaving orphaned units. The
added tests ensure that those orphans are removed by a call to the
orphan remove API endpoint.

Fix #134

### CODE REVIEW

File: pulp_smash/tests/rpm/api_v2/test_orphan_remove.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    def setUpClass(cls):
+        """Create an RPM repo, sync it, delete the repo, remove orphans."""
+        super(OrphanRemoveAllTestCase, cls).setUpClass()
+        client = api.Client(cls.cfg, api.json_handler)
+        body = _gen_repo()
+        body['importer_config']['feed'] = RPM_FEED_URL
+        repo = client.post(REPOSITORY_PATH, body)
+        client.response_handler = api.echo_handler
+        sync_path = urljoin(repo['_href'], 'actions/sync/')
+        cls.report = client.post(sync_path, {'override_config': {}})
```
Comment (Ichimonji10) [inline pulp_smash/tests/rpm/api_v2/test_orphan_remove.py]: I see we're making `report` a class attribute here, but we're not actually using this attribute in any other methods.

Comment (Ichimonji10) [inline pulp_smash/tests/rpm/api_v2/test_orphan_remove.py]: Same question about `cls.tasks`.

File: pulp_smash/tests/rpm/api_v2/test_orphan_remove.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    @classmethod
+    def setUpClass(cls):
+        """Create an RPM repo, sync it, delete the repo, remove orphans."""
+        super(OrphanRemoveAllTestCase, cls).setUpClass()
+        client = api.Client(cls.cfg, api.json_handler)
+        body = _gen_repo()
+        body['importer_config']['feed'] = RPM_FEED_URL
+        repo = client.post(REPOSITORY_PATH, body)
+        client.response_handler = api.echo_handler
```
Comment (Ichimonji10) [inline pulp_smash/tests/rpm/api_v2/test_orphan_remove.py]: What's the reasoning behind making the response handler `api.echo_handler` here?

If the response handler is left to be `api.json_handler`, or if it's set to `api.safe_handler`, then the calls to `cls.report.raise_for_status` and `api.poll_spawned_tasks()`, below, can be dropped. Not only that, but the `safe_handler` and `json_handler` make sure to check the contents of each completed task to ensure there are no errors or tracebacks. Handy!

File: pulp_smash/tests/rpm/api_v2/test_orphan_remove.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
+# coding=utf-8
+"""Test the removal of rpm orphans.
+
+This module assumes that the tests in
+:mod:`pulp_smash.tests.platform.api_v2.test_repository` and
++`pulp_smash.tests.platform.api_v2.test_sync_publish` hold true.
+"""
+
```
Comment (Ichimonji10) [inline pulp_smash/tests/rpm/api_v2/test_orphan_remove.py]: Can we add a `from __future__ import unicode_literals` here? It's a nice little way to prevent errors.

**Comment by Ichimonji10** [comment]
This looks great. I'd normally just make these changes myself after, but since you asked for the comments, there you go. :smile:

File: pulp_smash/tests/rpm/api_v2/test_orphan_remove.py (hunk 2 above)
Comment (asmacdo) [inline pulp_smash/tests/rpm/api_v2/test_orphan_remove.py]: No reason for these other than borrowing from other tests. Will be removed in update.

**Comment by asmacdo** [comment]
@Ichimonji10, I made the suggested changes. For what it's worth, I thought all of them were reasonable to expect contributors to fix.

I am getting some seemingly unrelated failures on my own box, http://fpaste.org/332345/94276914/

**Comment by Ichimonji10** [comment]
Thank you!

For what it's worth, I'm not receiving those errors on the systems I'm testing against, which are freshly provisioned Fedora 23 systems.

I'm getting a test failure when running these tests against Pulp 2.7:

```
(pulp-smash3) [ichimonji10@beech:pulp-smash]$ PULP_SMASH_CONFIG_FILE=${PULP}.json python -m unittest2 pulp_smash.tests.rpm.api_v2.test_orphan_remove
F.
======================================================================
FAIL: test_orphans_created (pulp_smash.tests.rpm.api_v2.test_orphan_remove.OrphanRemoveAllTestCase)
Ensure that orphans were created by deleting the repository.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ichimonji10/code/pulp-smash/pulp_smash/tests/rpm/api_v2/test_orphan_remove.py", line 60, in test_orphans_created
    self.assertGreaterEqual(self.num_orphans_post_repo_del, 39)
AssertionError: 3 not greater than or equal to 39

----------------------------------------------------------------------
Ran 2 tests in 22.395s

FAILED (failures=1)
```

Is this expected for Pulp 2.7? (Pro tip: using `assertGreaterEqual` produces a more useful error message than `assertTrue`)

File: pulp_smash/tests/rpm/api_v2/test_orphan_remove.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+from pulp_smash import api, utils
+from pulp_smash.constants import (
+    ORPHANS_PATH,
+    REPOSITORY_PATH,
+    RPM_FEED_URL,
+)
+
+
+def _gen_repo():
```
Comment (Ichimonji10) [inline pulp_smash/tests/rpm/api_v2/test_orphan_remove.py]: FYI, `pulp_smash.tests.rpm.api_v2.utils` has a function `gen_repo`. The above can become:

```python
from pulp_smash import api, utils
from pulp_smash.constants import ORPHANS_PATH, REPOSITORY_PATH, RPM_FEED_URL
from pulp_smash.tests.rpm.api_v2.utils import gen_repo
```

File: pulp_smash/tests/rpm/api_v2/test_orphan_remove.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    def setUpClass(cls):
+        """Create an RPM repo, sync it, delete the repo, remove orphans."""
+        super(OrphanRemoveAllTestCase, cls).setUpClass()
+        client = api.Client(cls.cfg, api.json_handler)
+        body = _gen_repo()
+        body['importer_config']['feed'] = RPM_FEED_URL
+        repo = client.post(REPOSITORY_PATH, body)
+        sync_path = urljoin(repo['_href'], 'actions/sync/')
+        client.post(sync_path, {'override_config': {}})
+        client.response_handler = api.json_handler
```
Comment (Ichimonji10) [inline pulp_smash/tests/rpm/api_v2/test_orphan_remove.py]: FYI, this line is a no-op. It can be dropped without any effect. (I'll do this in a follow-up commit if you don't have time.)

**Comment by asmacdo** [comment]
That test is a sanity check. It isn't really a failure, but it is a red flag to suggest that something else isn't working as expected. Perhaps you have another repo in the db that shares some content with this one?

File: pulp_smash/tests/rpm/api_v2/test_orphan_remove.py (hunk 5 above)
Comment (asmacdo) [inline pulp_smash/tests/rpm/api_v2/test_orphan_remove.py]: I see, I wasn't familiar with the Client class.

File: pulp_smash/tests/rpm/api_v2/test_orphan_remove.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
+# coding=utf-8
+"""Test the removal of rpm orphans.
+
+This module assumes that the tests in
+:mod:`pulp_smash.tests.platform.api_v2.test_repository` and
++`pulp_smash.tests.platform.api_v2.test_sync_publish` hold true.
```
Comment (Ichimonji10) [inline pulp_smash/tests/rpm/api_v2/test_orphan_remove.py]: This line is syntactically and semantically broken. But the automatic linters don't warn you, because this module isn't currently included in the documentation. This is what it takes to include it:

```diff
diff --git a/docs/api.rst b/docs/api.rst
index 1be39bd..79a7886 100644
--- a/docs/api.rst
+++ b/docs/api.rst
@@ -44,6 +44,7 @@ developers, not a gospel.
     api/pulp_smash.tests.rpm.api_v2.test_broker
     api/pulp_smash.tests.rpm.api_v2.test_duplicate_uploads
     api/pulp_smash.tests.rpm.api_v2.test_iso_crud
+    api/pulp_smash.tests.rpm.api_v2.test_orphan_remove
     api/pulp_smash.tests.rpm.api_v2.test_remove_unit
     api/pulp_smash.tests.rpm.api_v2.test_retain_old_count
     api/pulp_smash.tests.rpm.api_v2.test_sync_publish
diff --git a/docs/api/pulp_smash.tests.rpm.api_v2.test_orphan_remove.rst b/docs/api/pulp_smash.tests.rpm.api_v2.test_orphan_remove.rst
new file mode 100644
index 0000000..ad35039
--- /dev/null
+++ b/docs/api/pulp_smash.tests.rpm.api_v2.test_orphan_remove.rst
@@ -0,0 +1,7 @@
+`pulp_smash.tests.rpm.api_v2.test_orphan_remove`
+================================================
+
+Location: :doc:`/index` → :doc:`/api` →
+:doc:`/api/pulp_smash.tests.rpm.api_v2.test_orphan_remove`
+
+.. automodule:: pulp_smash.tests.rpm.api_v2.test_orphan_remove
```

Yes, this is a pain in the butt, and I'd like to have documentation stubs generated automatically. See #18. (I'll do this in a follow-up commit if you don't have time.)

File: pulp_smash/tests/rpm/api_v2/test_orphan_remove.py (hunk 5 above)
Comment (Ichimonji10) [inline pulp_smash/tests/rpm/api_v2/test_orphan_remove.py]: To be more precise, this line does have an effect. But the response handler is set earlier, so setting it is equivalent to:
```python
x = 1
y = x + 1
x = 1
```

File: pulp_smash/tests/rpm/api_v2/test_orphan_remove.py (hunk 6 above)
Comment (asmacdo) [inline pulp_smash/tests/rpm/api_v2/test_orphan_remove.py]: I'll leave this one for you.

Comment (Ichimonji10) [inline pulp_smash/tests/rpm/api_v2/test_orphan_remove.py]: Sure thing. :+1:

**Comment by asmacdo** [comment]
@Ichimonji10, did you figure out the 2.7 issue?

Also, how would you like to do this merge? Want me to do it or do you plan to do your cleanup at the same time?

**Comment by Ichimonji10** [comment]
I haven't yet figured out why the test fails for Pulp 2.7.

How to merge? Unless you'd like to do something different, I'm going to rebase the commit in this PR onto master, add a second commit with several changes, and merge into that. It'll produce a straight commit history with no merge commits. It currently looks like so:

scrot

See https://github.com/PulpQE/pulp-smash/compare/master...Ichimonji10:145

**Comment by asmacdo** [comment]
> How to merge? Unless you'd like to do something different, I'm going to rebase the commit in this PR onto master, add a second commit with several changes, and merge into that.
:+1: 

I am a little concerned about the 2.7 issue though. I don't think orphan remove was broken there.

**Comment by asmacdo** [comment]
@Ichimonji10, you have a 2.7 env handy though, a quick sanity check is to create a fresh db, create and sync zoo repo, delete zoo repo, and then `pulp-admin orphan list`

**Comment by Ichimonji10** [comment]
Aaaah, sorry! I'm getting this totally backward. The test fails on Pulp dev (2.8), and completely passes on 2.7.

**Comment by Ichimonji10** [comment]
Adding a call to `reset_pulp` at the beginning of the test case allows it to pass on Pulp 2.8.

**Comment by asmacdo** [comment]
> Adding a call to reset_pulp at the beginning of the test case allows it to pass on Pulp 2.8.
Ok, cool, seems like the thing to do. But that does indicate that some tests are not cleaning up after themselves.

**Comment by Ichimonji10** [comment]
Rut row. I'll keep an eye out for that.

**Comment by asmacdo** [comment]
@Ichimonji10, a clue: The offending test almost certainly syncs the zoo repo.

**Comment by Ichimonji10** [comment]
:smile: Thank you.

**Comment by Ichimonji10** [comment]
tl;dr The new tests are fine.

Sorry about all the noise regarding test failures. I've figured it out, and it's my own issue.

What happened is that I ran `pulp_smash.tests.rpm.api_v2.test_sync_publish.SyncOnDemandTestCase`, and the test failed to complete due to Squid redirecting my API client to an inaccessible hostname (apparently OS1 hosts are only accessible via IP, even when on the VPN?). Those tests are only compatible with Pulp 2.8, so I only encountered the consequences on Pulp 2.8.

**Comment by Ichimonji10** [comment]
Merged as 8c49cfaaf880afe04c815e5f1a9a578ebea068e5, with a follow-up commit of 0811270f995d686087caeb7ad9cee0c345820986. Thank you, @asmacdo!
